specs/api_cron_flow.md

```diff
@@ -0,0 +1,130 @@
+the API in a short time, our access may be throttled or suspended. We
+need to be able to update the various external APIs in a reasonable
+fashion.
+
+## Background and Significance
+
+There are currently three APIs that we use to enhance the data sets
+from the SBA.
+
+The first is an API to get the Yelp rating where possible.
+
+The second is a Google Civics API to get the Congressional district
+based on the borrower address.
+
+The third is an attempt to obtain the latitude and longitude for the
+borrower addresses.
+
+Given the size of our database (even just for the SFDO loan set), if
+we try to update all the ratings in a single run, the API providers
+may throttle our access. The Google Civics API has a daily limit of
+25,000 calls and one of the potential goecode APIs has a 2,500 daily
+limit.
+
+The request is to be able to chunk the output into small batches and
+schedule a cron job to run repeatedly until all records have been
+attempted.
+
+## Final Result
+
+For this project, we will make changes to the script to allow it to
+run a specific set of records and then exit. We will provide the
+capability to keep track of which records have been attempted and
+which have not. We will provide the ability to update the external
+APIs separately based on different limitations. We will also provide
+an example of a cron scheduling job that would permit the entire
+database to be refreshed over a period of hours or days.
+
+## Proposed Approach
+
+### General Discussion of Approach
+
+We discussed an approach whereby a wrapper script is used to setup the
+database and Python environment. The script will then pass arguments
+down to the control script, which will select records to be processed
+by each API. The control script will pass a selection of sba_sfdo_id
+fields (or other appropriate key) to the execution script for each API
+that is to be run.
+
+This approach means that the API execution script does not need to
+worry about processing the arguments or filtering the records.  Other
+than updating the timestamp, the execution code has very little
+knowledge of the batch control process.
+
+The cron job will call the wrapper script with the appropriate
+args. There will be a number of entries in the cron file to properly
+process all the external APIs in a timely manner.
+
+### Updates to the API control script
+
+The API control script will support the following new arguments.
+
+At least one of the following must be specified (all three may be
+specified if desired):
+
+--yelp: indicates that the script should update the Yelp entries.
+
+--civics: indicates that the script should update the Congressional
+  district entries.
+
+--geocode: indicates tha the script should update the Geocode
+  entries. Optional, if not specified these records are unchanged.
+
+The following are optional but recomended:
+
+--max_attempts <integer>: indicates the maximum number of records to
```

Comment:
Is there a way to set a default value that is equal to the maximum allowed by the API? For example if Google is 2,500; but I think Yelp is 50,000 what's the best way for us to set as default "for google should only try 2,500 rows, but for Yelp can go up to 50,000?

Reply:
I think we can set the default. I suggest that the section of the script which handles a specific external API can set a default, which the user can override (but perhaps not exceed).